django/db/migrations/state.py

```diff
@@ -109,6 +112,12 @@ def remove_model(self, app_label, model_name):
             # Need to do this explicitly since unregister_model() doesn't clear
             # the cache automatically (#24513)
             self.apps.clear_cache()
+            self.relations.pop((app_label, model_name), '')
+            relations = copy.copy(self.relations)
+            for key, relation in relations.items():
+                self.relations[key].pop((app_label, model_name), '')
+                if relation == defaultdict(list):
+                    self.relations.pop(key)
```

Comment:
This change needs to be performed even if `apps` is not cached as that's an independent registry.

A few things seems wrong/unnecessary here. I don't think there's anything wrong in leaving 

```suggestion
            removed_model_key = (app_label, model_name)
            self.relations.pop(removed_model_key, None)
            for model_key, model_relations in list(relations.items()):
                model_relations.pop(removed_model_key, None)
                if not model_relations:
                    self.relations.pop(model_key)
```